# 데이터 사전 처리 
 - 머신러닝 등 데이터 분석의 정확도는 분석 데이터의 품질에 좌우된다.
 - 데이터의 품질을 높이기 위해서는 누락, 중복 데이터 등 오류를 수정하고 분석 목적에 맞게 변형하는 과정이 필요하다.
 - 수집한 데이터를 분석 목적에 적합하도록 하는 작업을 사전처리 라고 한다. 

# 누락 데이터(NaN: Not a Number) 
 - 데이터를 파일로 입력할 때 빠트리거나 파일 형식을 변환하는 과정에서 데이터가 손실되는 경우가 발생 
 - 머신러닝 분석 모형에 데이터를 입력하기 전에 반드시 누락 데이터를 제거하고 다른 적절한 값으로 대체 하여야 한다. 
 - 누락데이터가 많이자면 데이터의 품질이 떨어지고 머신러닝 분석 알고리즘을 왜곡하는 현상이 발생 

# 누락데이터 확인

In [1]:
import seaborn as sns
import pandas as pd
import numpy as np 
import warnings 
warnings.filterwarnings('ignore')

In [2]:
# Titanic Dataset 가져오기
df = sns.load_dataset('titanic')
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


> deck에는 203개의 유효한 범주형 데이터만 있고 나머지 688개가 NaN이다. 

In [8]:
# deck 열의 NaN 갯수 계산하기 
nan_deck = df['deck'].value_counts(dropna = False)
nan_deck

NaN    688
C       59
B       47
D       33
E       32
A       15
F       13
G        4
Name: deck, dtype: int64

# 누락 데이터 찾기

In [11]:
df.head().isnull() # 누락이면 True 발생 

survived  pclass    sex    age  sibsp  parch   fare  embarked  class  \
0     False   False  False  False  False  False  False     False  False   
1     False   False  False  False  False  False  False     False  False   
2     False   False  False  False  False  False  False     False  False   
3     False   False  False  False  False  False  False     False  False   
4     False   False  False  False  False  False  False     False  False   

     who  adult_male   deck  embark_town  alive  alone  
0  False       False   True        False  False  False  
1  False       False  False        False  False  False  
2  False       False   True        False  False  False  
3  False       False  False        False  False  False  
4  False       False   True        False  False  False

In [12]:
df.head().notnull() # 누락이면 False

survived  pclass   sex   age  sibsp  parch  fare  embarked  class   who  \
0      True    True  True  True   True   True  True      True   True  True   
1      True    True  True  True   True   True  True      True   True  True   
2      True    True  True  True   True   True  True      True   True  True   
3      True    True  True  True   True   True  True      True   True  True   
4      True    True  True  True   True   True  True      True   True  True   

   adult_male   deck  embark_town  alive  alone  
0        True  False         True   True   True  
1        True   True         True   True   True  
2        True  False         True   True   True  
3        True   True         True   True   True  
4        True  False         True   True   True

In [13]:
# 누락데이터 갯수 확인 
df.head().isnull().sum(axis=0)

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
deck           3
embark_town    0
alive          0
alone          0
dtype: int64

# 누락데이터 제거 

In [16]:
# for 문으로 각 열의 NaN 개수 계산 
missing_df = df.isnull()
missing_df

survived  pclass    sex    age  sibsp  parch   fare  embarked  class  \
0       False   False  False  False  False  False  False     False  False   
1       False   False  False  False  False  False  False     False  False   
2       False   False  False  False  False  False  False     False  False   
3       False   False  False  False  False  False  False     False  False   
4       False   False  False  False  False  False  False     False  False   
..        ...     ...    ...    ...    ...    ...    ...       ...    ...   
886     False   False  False  False  False  False  False     False  False   
887     False   False  False  False  False  False  False     False  False   
888     False   False  False   True  False  False  False     False  False   
889     False   False  False  False  False  False  False     False  False   
890     False   False  False  False  False  False  False     False  False   

       who  adult_male   deck  embark_town  alive  alone  
0    False       False   True        False  False  False  
1    False       False  False        False  False  False  
2    False       False   True        False  False  False  
3    False       False  False        False  False  False  
4    False       False   True        False  False  False  
..     ...         ...    ...          ...    ...    ...  
886  False       False   True        False  False  False  
887  False       False  False        False  False  False  
888  False       False   True        False  False  False  
889  False       False  False        False  False  False  
890  False       False   True        False  False  False  

[891 rows x 15 columns]

In [19]:
for col in missing_df.columns:
    missing_count = missing_df[col].value_counts()
    try:
        print(col,":",missing_count[True]) # NaN 값이 있으면 개수를 출력 
    except:
        print(col,":",0) # NaN 값이 없으면 0을 출력 

survived : 0
pclass : 0
sex : 0
age : 177
sibsp : 0
parch : 0
fare : 0
embarked : 2
class : 0
who : 0
adult_male : 0
deck : 688
embark_town : 2
alive : 0
alone : 0


In [20]:
# NaN의 갯수가 500개 이상인 열을 모두 삭제 
df_thresh = df.dropna(axis=1, thresh=500)
df_thresh.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'embark_town', 'alive',
       'alone'],
      dtype='object')

In [22]:
# age열에 나이 데이터가 없는 모든 행을 삭제 
df_thresh_age = df_thresh.dropna(subset=['age'], how='any', axis=0)
len(df_thresh_age)

714

In [23]:
df_thresh_age.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male  embark_town alive  alone  
0    man        True  Southampton    no  False  
1  woman       False    Cherbourg   yes  False  
2  woman       False  Southampton   yes   True  
3  woman       False  Southampton   yes  False  
4    man        True  Southampton    no   True

---
# 누락 데이터의 치환
 - 누락 데이터 무작성 삭제해 버린다면 어렵게 수집한 데이터를 활용허지 못한다. 
 - 데이터 중에서 일부가 누락되어 있더라도 나머지 데이터를 최대한 살려서 데이터 분석에 활용하는 경우 좋은 결과를 도출 할 수 있다.

In [24]:
df = sns.load_dataset('titanic')
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [25]:
# age열의 첫 10개 데이터 출력 
df['age'].head(10)

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
5     NaN
6    54.0
7     2.0
8    27.0
9    14.0
Name: age, dtype: float64

In [28]:
# age 열의 NaN 값을 다른 나이 데이터의 평균으로 변경하기 
mean_age = df['age'].mean()
df['age'].fillna(mean_age, inplace=True)
df['age'].head(10)

0    22.000000
1    38.000000
2    26.000000
3    35.000000
4    35.000000
5    29.699118
6    54.000000
7     2.000000
8    27.000000
9    14.000000
Name: age, dtype: float64

In [31]:
# embark_town열의 829행의 NaN 데이터 출력 
df['embark_town'][825:830]

825     Queenstown
826    Southampton
827      Cherbourg
828     Queenstown
829            NaN
Name: embark_town, dtype: object

In [33]:
# 빈도수가 높을 것 추출 
most_freq = df['embark_town'].value_counts(dropna=True).idxmax()
most_freq

'Southampton'

In [34]:
#NaN 변경 
df['embark_town'].fillna(most_freq,inplace=True)
df['embark_town'][825:830]

825     Queenstown
826    Southampton
827      Cherbourg
828     Queenstown
829    Southampton
Name: embark_town, dtype: object

--- 
# 누락 데이터가 NaN으로 표시되지 않는 경우 
 - 데이터셋중에서 누락 데이터가 NaN으로 입력되지 않는 경우도 많다. 예를 틀면 숫자 0이나 문자 -,? 같은 값으로 입력된다.

import numpy as np           
df.replace('-',np.nan,inplace=True)

# 앞뒤의 유사성 데이터로 값을 치환

In [35]:
df = sns.load_dataset('titanic')
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [36]:
df['embark_town'][825:830]

825     Queenstown
826    Southampton
827      Cherbourg
828     Queenstown
829            NaN
Name: embark_town, dtype: object

In [37]:
# embark_town 열의 NaN 값을 바로 앞에 있는 열 값으로 변경하기 
df['embark_town'].fillna(method='ffill', inplace=True)
df['embark_town'][825:830]


825     Queenstown
826    Southampton
827      Cherbourg
828     Queenstown
829     Queenstown
Name: embark_town, dtype: object

---
# 중복 데이터 처리 
: 동일한 대상이 중복으로 존재하는 것은 분석결과의 왜곡을 초래한다.

In [38]:
### 중복 데이터 확인
df = pd.DataFrame({'c1':['a','a','b','a','b '],
                   'c2':[1,1,1,2,2],
                   'c3':[1,1,2,2,2]
                  })
df

c1  c2  c3
0   a   1   1
1   a   1   1
2   b   1   2
3   a   2   2
4  b    2   2

In [40]:
# 데이터프레임 전체 행 데이터 중에서 중복값 찾기 
df_dup = df.duplicated()
df_dup

0    False
1     True
2    False
3    False
4    False
dtype: bool

In [41]:
# 데이터 프레임 특정 열 데이터에서 중복값 찾기 
col_dup = df['c2'].duplicated()
col_dup

0    False
1     True
2     True
3    False
4     True
Name: c2, dtype: bool

## 중복 데이터 제거 

In [42]:
# 데이터프레임에서 중복 행 제거 
df2 = df.drop_duplicates()
df2

c1  c2  c3
0   a   1   1
2   b   1   2
3   a   2   2
4  b    2   2

In [43]:
# C2, C3열을 기준으로 중복 행 제거 
df3 = df.drop_duplicates(subset=['c2','c3'])
df3

c1  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2

---
# 데이터 표준화 
 - 여러곳에서 수집한 데이터는 서로 단위가 섞여 있거나 대상을 다른 형식으로 표현하는 경우가 많아졌다. 
 - 대상을 표현하는 방법에 차이가 있으면 분석의 정확도는 현저히 낮아진다. 
 - 데이터 포맷을 일관성 있게 하는 작업이 표준화 이다. 
    

# 단위 환산

In [45]:
df = pd.read_csv("./Data/auto-mpg.csv", header=None)
df.head()

0  1      2      3       4     5   6  7                          8
0  18.0  8  307.0  130.0  3504.0  12.0  70  1  chevrolet chevelle malibu
1  15.0  8  350.0  165.0  3693.0  11.5  70  1          buick skylark 320
2  18.0  8  318.0  150.0  3436.0  11.0  70  1         plymouth satellite
3  16.0  8  304.0  150.0  3433.0  12.0  70  1              amc rebel sst
4  17.0  8  302.0  140.0  3449.0  10.5  70  1                ford torino

In [46]:
df.columns = ['mpg','cylinders','displacement','horsepower','weight',
              'acceleration','model year','origin','name']

In [47]:
df.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0      130.0  3504.0          12.0          70   
1  15.0          8         350.0      165.0  3693.0          11.5          70   
2  18.0          8         318.0      150.0  3436.0          11.0          70   
3  16.0          8         304.0      150.0  3433.0          12.0          70   
4  17.0          8         302.0      140.0  3449.0          10.5          70   

   origin                       name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

In [48]:
# mpg(mile per gallon) -> kpl(kilometer per liter)로 변환 
# 1 mile = 1.60934km 
# 1 gallon = 3.78541 liter
mpg_to_kpl = 1.60934 / 3.78541
mpg_to_kpl


0.42514285110463595

In [ ]:
# kpl이라는 열을 추가 하고 mpg * mpg_to_kpl의 값을 추가하기 

In [50]:
df['kpl'] = df['mpg'] * mpg_to_kpl
df

mpg  cylinders  displacement horsepower  weight  acceleration  \
0    18.0          8         307.0      130.0  3504.0          12.0   
1    15.0          8         350.0      165.0  3693.0          11.5   
2    18.0          8         318.0      150.0  3436.0          11.0   
3    16.0          8         304.0      150.0  3433.0          12.0   
4    17.0          8         302.0      140.0  3449.0          10.5   
..    ...        ...           ...        ...     ...           ...   
393  27.0          4         140.0      86.00  2790.0          15.6   
394  44.0          4          97.0      52.00  2130.0          24.6   
395  32.0          4         135.0      84.00  2295.0          11.6   
396  28.0          4         120.0      79.00  2625.0          18.6   
397  31.0          4         119.0      82.00  2720.0          19.4   

     model year  origin                       name        kpl  
0            70       1  chevrolet chevelle malibu   7.652571  
1            70       1          buick skylark 320   6.377143  
2            70       1         plymouth satellite   7.652571  
3            70       1              amc rebel sst   6.802286  
4            70       1                ford torino   7.227428  
..          ...     ...                        ...        ...  
393          82       1            ford mustang gl  11.478857  
394          82       2                  vw pickup  18.706285  
395          82       1              dodge rampage  13.604571  
396          82       1                ford ranger  11.904000  
397          82       1                 chevy s-10  13.179428  

[398 rows x 10 columns]

In [51]:
# kpl 열을 소주점 둘째자리 보이기 
df['kpl'] = df['kpl'].round(2)
df.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0      130.0  3504.0          12.0          70   
1  15.0          8         350.0      165.0  3693.0          11.5          70   
2  18.0          8         318.0      150.0  3436.0          11.0          70   
3  16.0          8         304.0      150.0  3433.0          12.0          70   
4  17.0          8         302.0      140.0  3449.0          10.5          70   

   origin                       name   kpl  
0       1  chevrolet chevelle malibu  7.65  
1       1          buick skylark 320  6.38  
2       1         plymouth satellite  7.65  
3       1              amc rebel sst  6.80  
4       1                ford torino  7.23

# 자료형 변환 

In [52]:
# 각 열의 자료형 확인 
df.dtypes

mpg             float64
cylinders         int64
displacement    float64
horsepower       object
weight          float64
acceleration    float64
model year        int64
origin            int64
name             object
kpl             float64
dtype: object

> horesepower가 object 자료형이다 <-- 숫자외의 다른 자료형이 들어가 있다.

In [54]:
# horsepower 열의 고윳값 확인 
df['horsepower'].unique()

array(['130.0', '165.0', '150.0', '140.0', '198.0', '220.0', '215.0',
       '225.0', '190.0', '170.0', '160.0', '95.00', '97.00', '85.00',
       '88.00', '46.00', '87.00', '90.00', '113.0', '200.0', '210.0',
       '193.0', '?', '100.0', '105.0', '175.0', '153.0', '180.0', '110.0',
       '72.00', '86.00', '70.00', '76.00', '65.00', '69.00', '60.00',
       '80.00', '54.00', '208.0', '155.0', '112.0', '92.00', '145.0',
       '137.0', '158.0', '167.0', '94.00', '107.0', '230.0', '49.00',
       '75.00', '91.00', '122.0', '67.00', '83.00', '78.00', '52.00',
       '61.00', '93.00', '148.0', '129.0', '96.00', '71.00', '98.00',
       '115.0', '53.00', '81.00', '79.00', '120.0', '152.0', '102.0',
       '108.0', '68.00', '58.00', '149.0', '89.00', '63.00', '48.00',
       '66.00', '139.0', '103.0', '125.0', '133.0', '138.0', '135.0',
       '142.0', '77.00', '62.00', '132.0', '84.00', '64.00', '74.00',
       '116.0', '82.00'], dtype=object)

> 데이터의 중간에 ? 가 있음

In [55]:
# Data의 갯수 확인 
len(df['horsepower'])

398

In [57]:
# Data가 ? 확인 
mask = df['horsepower'] == '?'
mask

0      False
1      False
2      False
3      False
4      False
       ...  
393    False
394    False
395    False
396    False
397    False
Name: horsepower, Length: 398, dtype: bool

In [58]:
df[mask]

mpg  cylinders  displacement horsepower  weight  acceleration  \
32   25.0          4          98.0          ?  2046.0          19.0   
126  21.0          6         200.0          ?  2875.0          17.0   
330  40.9          4          85.0          ?  1835.0          17.3   
336  23.6          4         140.0          ?  2905.0          14.3   
354  34.5          4         100.0          ?  2320.0          15.8   
374  23.0          4         151.0          ?  3035.0          20.5   

     model year  origin                  name    kpl  
32           71       1            ford pinto  10.63  
126          74       1         ford maverick   8.93  
330          80       2  renault lecar deluxe  17.39  
336          80       1    ford mustang cobra  10.03  
354          81       2           renault 18i  14.67  
374          82       1        amc concord dl   9.78

In [59]:
len(df[mask])

6

In [60]:
# ?를 NaN으로 변경 
df['horsepower'].replace('?', np.nan, inplace=True)
df['horsepower'].unique()

array(['130.0', '165.0', '150.0', '140.0', '198.0', '220.0', '215.0',
       '225.0', '190.0', '170.0', '160.0', '95.00', '97.00', '85.00',
       '88.00', '46.00', '87.00', '90.00', '113.0', '200.0', '210.0',
       '193.0', nan, '100.0', '105.0', '175.0', '153.0', '180.0', '110.0',
       '72.00', '86.00', '70.00', '76.00', '65.00', '69.00', '60.00',
       '80.00', '54.00', '208.0', '155.0', '112.0', '92.00', '145.0',
       '137.0', '158.0', '167.0', '94.00', '107.0', '230.0', '49.00',
       '75.00', '91.00', '122.0', '67.00', '83.00', '78.00', '52.00',
       '61.00', '93.00', '148.0', '129.0', '96.00', '71.00', '98.00',
       '115.0', '53.00', '81.00', '79.00', '120.0', '152.0', '102.0',
       '108.0', '68.00', '58.00', '149.0', '89.00', '63.00', '48.00',
       '66.00', '139.0', '103.0', '125.0', '133.0', '138.0', '135.0',
       '142.0', '77.00', '62.00', '132.0', '84.00', '64.00', '74.00',
       '116.0', '82.00'], dtype=object)

In [61]:
# 누락데이터 행 삭제 
df.dropna(subset=['horsepower'], axis=0, inplace=True)
len(df['horsepower'])

392

In [63]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 392 entries, 0 to 397
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           392 non-null    float64
 1   cylinders     392 non-null    int64  
 2   displacement  392 non-null    float64
 3   horsepower    392 non-null    object 
 4   weight        392 non-null    float64
 5   acceleration  392 non-null    float64
 6   model year    392 non-null    int64  
 7   origin        392 non-null    int64  
 8   name          392 non-null    object 
 9   kpl           392 non-null    float64
dtypes: float64(5), int64(3), object(2)
memory usage: 33.7+ KB


In [64]:
# 문자열을 실수형으로 변환 
df['horsepower'] = df['horsepower'].astype('float')
df['horsepower'].dtypes

dtype('float64')

In [65]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 392 entries, 0 to 397
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           392 non-null    float64
 1   cylinders     392 non-null    int64  
 2   displacement  392 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        392 non-null    float64
 5   acceleration  392 non-null    float64
 6   model year    392 non-null    int64  
 7   origin        392 non-null    int64  
 8   name          392 non-null    object 
 9   kpl           392 non-null    float64
dtypes: float64(6), int64(3), object(1)
memory usage: 33.7+ KB


In [67]:
# origin을 국가명으로 변경 
df['origin'].unique()

array([1, 3, 2])

In [69]:
df['origin'].dtypes

dtype('int64')

In [70]:
df['origin'].replace({1:'USA', 2:'EU', 3:'JAPAN'}, inplace=True)
df.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0  3504.0          12.0   
1  15.0          8         350.0       165.0  3693.0          11.5   
2  18.0          8         318.0       150.0  3436.0          11.0   
3  16.0          8         304.0       150.0  3433.0          12.0   
4  17.0          8         302.0       140.0  3449.0          10.5   

   model year origin                       name   kpl  
0          70    USA  chevrolet chevelle malibu  7.65  
1          70    USA          buick skylark 320  6.38  
2          70    USA         plymouth satellite  7.65  
3          70    USA              amc rebel sst  6.80  
4          70    USA                ford torino  7.23

In [72]:
# origin 열의 문자열 자료형을 범주형으로 변환 
df['origin'] = df['origin'].astype('category')
df['origin'].dtypes

CategoricalDtype(categories=['EU', 'JAPAN', 'USA'], ordered=False)

In [73]:
# 범주형을 문자열로 다시 변환 
df['origin'] = df['origin'].astype('str')
df['origin'].dtypes

dtype('O')

In [78]:
# model year가 정수로 되어 있으나 수치계산과 관련 없으므로 범주형으로 변환 
df['model year'].sample(3)

270    78
100    73
221    77
Name: model year, dtype: category
Categories (13, int64): [70, 71, 72, 73, ..., 79, 80, 81, 82]

In [77]:
df['model year'] = df['model year'].astype('category')

--
# 범주형(카테고리) 데이터 처리 
 - 연속 데이터를 그대로 사용하기 보다는 일정한 구간(binning)으로 나눠서 분석하는 것이 효율적인 경우가 많다. 

In [80]:
count, bin_dividers = np.histogram(df['horsepower'], bins=3)
bin_dividers

array([ 46.        , 107.33333333, 168.66666667, 230.        ])

> 46 ~ 107.33333333, 107.3333333333 ~ 168.666666667, 168.666666667 ~ 230의 3구간

In [82]:
# 3개의 bin에 이름을 지정 
bin_names = ['저출력', '보통출력', '고출력']

In [83]:
df['hp_bin'] = pd.cut(x=df['horsepower'], # 데이터 값 배열
                      bins = bin_dividers, # 경계 값 리스트
                      labels = bin_names, # bin 이름 
                      include_lowest=True # 첫 경계값 포함 
                    )

In [84]:
df[['horsepower','hp_bin']].head(15)

horsepower hp_bin
0        130.0   보통출력
1        165.0   보통출력
2        150.0   보통출력
3        150.0   보통출력
4        140.0   보통출력
5        198.0    고출력
6        220.0    고출력
7        215.0    고출력
8        225.0    고출력
9        190.0    고출력
10       170.0    고출력
11       160.0   보통출력
12       150.0   보통출력
13       225.0    고출력
14        95.0    저출력

---
## 더미(Dummy) 변수
- 범주형 데이터를 회귀분석이나 머신러닝에 바로 사용할 수 없는 경우 사용.
- 0 또는 1로 표현

In [85]:
horsepower_dummies = pd.get_dummies(df['hp_bin'])
horsepower_dummies.head(15)

저출력  보통출력  고출력
0     0     1    0
1     0     1    0
2     0     1    0
3     0     1    0
4     0     1    0
5     0     0    1
6     0     0    1
7     0     0    1
8     0     0    1
9     0     0    1
10    0     0    1
11    0     1    0
12    0     1    0
13    0     0    1
14    1     0    0

--- 
#One-Hot Encoding 
 - 범주형 데이터를 0,1을 갖는 원핫벡터로 변환 
 - 결과 값은 선형대수학에서 정의하는 희소행렬로 변환한다. 
 - 희소행렬은 좌표와 값 형태로 정리 

In [86]:
# sklearn Library
from sklearn import preprocessing

In [87]:
# 전처리를 위한 encoder 만들기 
label_encoder = preprocessing.LabelEncoder()
onehot_encoder = preprocessing.OneHotEncoder()

In [90]:
# label encoder로 문자열 범주를 숫자형으로 변환 
onehot_labeled = label_encoder.fit_transform(df['hp_bin'].head(15))
print(onehot_labeled)
print(type(onehot_labeled))

[1 1 1 1 1 0 0 0 0 0 0 1 1 0 2]
<class 'numpy.ndarray'>


In [91]:
# 2차원 행렬로 형태 변경 
onehot_reshaped = onehot_labeled.reshape(len(onehot_labeled),1)
print(onehot_reshaped)
print(type(onehot_reshaped))

[[1]
 [1]
 [1]
 [1]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [1]
 [0]
 [2]]
<class 'numpy.ndarray'>


In [92]:
# 희소 행렬로 변환 
onehot_fitted = onehot_encoder.fit_transform(onehot_reshaped)
print(onehot_fitted)


  (0, 1)	1.0
  (1, 1)	1.0
  (2, 1)	1.0
  (3, 1)	1.0
  (4, 1)	1.0
  (5, 0)	1.0
  (6, 0)	1.0
  (7, 0)	1.0
  (8, 0)	1.0
  (9, 0)	1.0
  (10, 0)	1.0
  (11, 1)	1.0
  (12, 1)	1.0
  (13, 0)	1.0
  (14, 2)	1.0


---
# 정규화(Normalization)
 - 각 열에 있는 숫자 데이터의 상대적 크기 때문에 머신러닝 분석결과가 달라질 수 있다. 
 - 숫자 데이터의 상태적인 크기 차이를 제거 

### 최대값의 절대값을 이용

In [93]:
# horsepower 열의 통계 정보로 최댓값 확인 
df.horsepower.describe()

count    392.000000
mean     104.469388
std       38.491160
min       46.000000
25%       75.000000
50%       93.500000
75%      126.000000
max      230.000000
Name: horsepower, dtype: float64

In [94]:
# horsepower 열의 최대값의 절대값으로 모든 데이터를 나눠서 저장 
horsepower1 = df.horsepower / abs(df.horsepower.max())
horsepower1.head()

0    0.565217
1    0.717391
2    0.652174
3    0.652174
4    0.608696
Name: horsepower, dtype: float64

In [95]:
horsepower1.describe()

count    392.000000
mean       0.454215
std        0.167353
min        0.200000
25%        0.326087
50%        0.406522
75%        0.547826
max        1.000000
Name: horsepower, dtype: float64

### 각 열의 데이터 중에서 최대값과 최소값을 뺀 값으로 나누는 방법 

In [97]:
# 각 열에서 해당열의 최소값을 뺀 값을 분자로 하고 
min_x = df.horsepower - df.horsepower.min()

# 해당 열의 최대값 최소값의 차를 분모를 하는 수를 계산하면 가장 큰 값을 1이 된다. 
min_max = df.horsepower.max() - df.horsepower.min()
horsepower2 = min_x / min_max
horsepower2.describe()


count    392.000000
mean       0.317768
std        0.209191
min        0.000000
25%        0.157609
50%        0.258152
75%        0.434783
max        1.000000
Name: horsepower, dtype: float64

 ---
# 시계열 데이터 
 - 특정한 시점을 기록하는 Timestamp
 - 두 지점 사이의 일정한 기간을 나타내는 Period

In [102]:
df = pd.read_csv('./Data/stock-data.csv')
df.head()

Date  Close  Start   High    Low  Volume
0  2018-07-02  10100  10850  10900  10000  137977
1  2018-06-29  10700  10550  10900   9990  170253
2  2018-06-28  10400  10900  10950  10150  155769
3  2018-06-27  10900  10800  11050  10500  133548
4  2018-06-26  10800  10900  11000  10700   63039

In [103]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Date    20 non-null     object
 1   Close   20 non-null     int64 
 2   Start   20 non-null     int64 
 3   High    20 non-null     int64 
 4   Low     20 non-null     int64 
 5   Volume  20 non-null     int64 
dtypes: int64(5), object(1)
memory usage: 1.1+ KB


> Data 컬럼이 object이다. 

In [104]:
# 문자열 데이터를 Timestamp로 변환 
df['new_Date'] = pd.to_datetime(df['Date'])
df.head()

Date  Close  Start   High    Low  Volume   new_Date
0  2018-07-02  10100  10850  10900  10000  137977 2018-07-02
1  2018-06-29  10700  10550  10900   9990  170253 2018-06-29
2  2018-06-28  10400  10900  10950  10150  155769 2018-06-28
3  2018-06-27  10900  10800  11050  10500  133548 2018-06-27
4  2018-06-26  10800  10900  11000  10700   63039 2018-06-26

In [105]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   Date      20 non-null     object        
 1   Close     20 non-null     int64         
 2   Start     20 non-null     int64         
 3   High      20 non-null     int64         
 4   Low       20 non-null     int64         
 5   Volume    20 non-null     int64         
 6   new_Date  20 non-null     datetime64[ns]
dtypes: datetime64[ns](1), int64(5), object(1)
memory usage: 1.2+ KB


In [107]:
type(df['new_Date'][0])

pandas._libs.tslibs.timestamps.Timestamp

In [108]:
# 시계열 값으로 변환된 열을 새로운 행 인덱스로 지정
df.set_index('new_Date', inplace=True)
df.head()

Date  Close  Start   High    Low  Volume
new_Date                                                  
2018-07-02  2018-07-02  10100  10850  10900  10000  137977
2018-06-29  2018-06-29  10700  10550  10900   9990  170253
2018-06-28  2018-06-28  10400  10900  10950  10150  155769
2018-06-27  2018-06-27  10900  10800  11050  10500  133548
2018-06-26  2018-06-26  10800  10900  11000  10700   63039

In [109]:
# 기존 날짜열 삭제 
df.drop('Date', axis=1, inplace=True)

In [110]:
df

Close  Start   High    Low  Volume
new_Date                                      
2018-07-02  10100  10850  10900  10000  137977
2018-06-29  10700  10550  10900   9990  170253
2018-06-28  10400  10900  10950  10150  155769
2018-06-27  10900  10800  11050  10500  133548
2018-06-26  10800  10900  11000  10700   63039
2018-06-25  11150  11400  11450  11000   55519
2018-06-22  11300  11250  11450  10750  134805
2018-06-21  11200  11350  11750  11200  133002
2018-06-20  11550  11200  11600  10900  308596
2018-06-19  11300  11850  11950  11300  180656
2018-06-18  12000  13400  13400  12000  309787
2018-06-15  13400  13600  13600  12900  201376
2018-06-14  13450  13200  13700  13150  347451
2018-06-12  13200  12200  13300  12050  558148
2018-06-11  11950  12000  12250  11950   62293
2018-06-08  11950  11950  12200  11800   59258
2018-06-07  11950  12200  12300  11900   49088
2018-06-05  12150  11800  12250  11800   42485
2018-06-04  11900  11900  12200  11700   25171
2018-06-01  11900  11800  12100  11750   32062

In [111]:
# Timestamp를 period로 변환 
pr_day = ts_dates.to_period(freq='D')
print(pr_day)

NameError: name 'ts_dates' is not defined

In [ ]:
# Timestamp를 period로 변환 
pr_day = ts_dates.to_period(freq='D')
print(pr_day)

In [ ]:
# Timestamp를 period로 변환 
pr_month = ts_dates.to_period(freq='M')
print(pr_year)

In [ ]:
# Timestamp를 period로 변환 
pr_year = ts_dates.to_period(freq='A')
print(pr_year)

In [112]:
D(Day:1일)
W(Week:1주)
M(Month end:월말)
MS(Month begin:월초)
Q(Quarter end: 분기말)
QS(Quarter begin: 분기 초)
A(year end:연말)
AS(year begin:연초)
B(Business Day:휴일제외)
H(hour:1시간)
T(minute:1분)
S(second:1초)

SyntaxError: invalid syntax (<ipython-input-112-425f6d7b59cf>, line 1)

In [113]:
# Timestamp의 배열 만들기 - 월 간격, 월의 시작일 기준 
ts_ms = pd.date_range(start='2019-01-01',
                     end =None,
                      periods=6,
                      freq='MS',
                      tz='Asia/Seoul'
                     )
ts_ms


DatetimeIndex(['2019-01-01 00:00:00+09:00', '2019-02-01 00:00:00+09:00',
               '2019-03-01 00:00:00+09:00', '2019-04-01 00:00:00+09:00',
               '2019-05-01 00:00:00+09:00', '2019-06-01 00:00:00+09:00'],
              dtype='datetime64[ns, Asia/Seoul]', freq='MS')

In [114]:
# 2019-01-01부터 월말기준  마지막일을 기준으로 6개의 Timestamp 만들기 
ts_me = pd.date_range(start='2019-01-01',
                     end =None,
                      periods=6,
                      freq='M',
                      tz='Asia/Seoul'
                     )
ts_me

DatetimeIndex(['2019-01-31 00:00:00+09:00', '2019-02-28 00:00:00+09:00',
               '2019-03-31 00:00:00+09:00', '2019-04-30 00:00:00+09:00',
               '2019-05-31 00:00:00+09:00', '2019-06-30 00:00:00+09:00'],
              dtype='datetime64[ns, Asia/Seoul]', freq='M')

In [115]:
# 2019-01-01부터 3개월씩 마지막일을 기준으로 6개의 Timestamp 만들기 
ts_me = pd.date_range(start='2019-01-01',
                     end =None,
                      periods=6,
                      freq='3M',
                      tz='Asia/Seoul'
                     )
ts_me

DatetimeIndex(['2019-01-31 00:00:00+09:00', '2019-04-30 00:00:00+09:00',
               '2019-07-31 00:00:00+09:00', '2019-10-31 00:00:00+09:00',
               '2020-01-31 00:00:00+09:00', '2020-04-30 00:00:00+09:00'],
              dtype='datetime64[ns, Asia/Seoul]', freq='3M')

--- 
# 시계열 데이터 활용 

In [117]:
df = pd.read_csv('./Data/stock-data.csv')
df['new_Date'] = pd.to_datetime(df['Date'])
df.head()

Date  Close  Start   High    Low  Volume   new_Date
0  2018-07-02  10100  10850  10900  10000  137977 2018-07-02
1  2018-06-29  10700  10550  10900   9990  170253 2018-06-29
2  2018-06-28  10400  10900  10950  10150  155769 2018-06-28
3  2018-06-27  10900  10800  11050  10500  133548 2018-06-27
4  2018-06-26  10800  10900  11000  10700   63039 2018-06-26

In [118]:
# dt 속성을 이용 연 월 일 저장 
df['Year'] = df['new_Date'].dt.year
df['Month'] = df['new_Date'].dt.month
df['Day'] = df['new_Date'].dt.day
df.head()

Date  Close  Start   High    Low  Volume   new_Date  Year  Month  Day
0  2018-07-02  10100  10850  10900  10000  137977 2018-07-02  2018      7    2
1  2018-06-29  10700  10550  10900   9990  170253 2018-06-29  2018      6   29
2  2018-06-28  10400  10900  10950  10150  155769 2018-06-28  2018      6   28
3  2018-06-27  10900  10800  11050  10500  133548 2018-06-27  2018      6   27
4  2018-06-26  10800  10900  11000  10700   63039 2018-06-26  2018      6   26

In [119]:
# Timestamp를 Period로 변환히여 년월일 표기 
df['date_yr'] = df['new_Date'].dt.to_period(freq='A')
df['date_m'] = df['new_Date'].dt.to_period(freq='M')
df.head()


Date  Close  Start   High    Low  Volume   new_Date  Year  Month  \
0  2018-07-02  10100  10850  10900  10000  137977 2018-07-02  2018      7   
1  2018-06-29  10700  10550  10900   9990  170253 2018-06-29  2018      6   
2  2018-06-28  10400  10900  10950  10150  155769 2018-06-28  2018      6   
3  2018-06-27  10900  10800  11050  10500  133548 2018-06-27  2018      6   
4  2018-06-26  10800  10900  11000  10700   63039 2018-06-26  2018      6   

   Day date_yr   date_m  
0    2    2018  2018-07  
1   29    2018  2018-06  
2   28    2018  2018-06  
3   27    2018  2018-06  
4   26    2018  2018-06

In [120]:
# 원하는 열을 새로운 행 인덱스로 지정 
df.set_index('date_m', inplace=True)
df.head()

Date  Close  Start   High    Low  Volume   new_Date  Year  \
date_m                                                                     
2018-07  2018-07-02  10100  10850  10900  10000  137977 2018-07-02  2018   
2018-06  2018-06-29  10700  10550  10900   9990  170253 2018-06-29  2018   
2018-06  2018-06-28  10400  10900  10950  10150  155769 2018-06-28  2018   
2018-06  2018-06-27  10900  10800  11050  10500  133548 2018-06-27  2018   
2018-06  2018-06-26  10800  10900  11000  10700   63039 2018-06-26  2018   

         Month  Day date_yr  
date_m                       
2018-07      7    2    2018  
2018-06      6   29    2018  
2018-06      6   28    2018  
2018-06      6   27    2018  
2018-06      6   26    2018

In [123]:
# 날짜 index 활용 
df = pd.read_csv('./Data/stock-data.csv')
df['new_Date'] = pd.to_datetime(df['Date'])
df.set_index('new_Date', inplace=True)

In [125]:
df.head()

Date  Close  Start   High    Low  Volume
new_Date                                                  
2018-07-02  2018-07-02  10100  10850  10900  10000  137977
2018-06-29  2018-06-29  10700  10550  10900   9990  170253
2018-06-28  2018-06-28  10400  10900  10950  10150  155769
2018-06-27  2018-06-27  10900  10800  11050  10500  133548
2018-06-26  2018-06-26  10800  10900  11000  10700   63039

In [126]:
# 날짜 인덱스를 이용하여 데이터 선택하기 
df_y = df['2018']
df_y.head()

Date  Close  Start   High    Low  Volume
new_Date                                                  
2018-07-02  2018-07-02  10100  10850  10900  10000  137977
2018-06-29  2018-06-29  10700  10550  10900   9990  170253
2018-06-28  2018-06-28  10400  10900  10950  10150  155769
2018-06-27  2018-06-27  10900  10800  11050  10500  133548
2018-06-26  2018-06-26  10800  10900  11000  10700   63039